File: booking/src/main/java/learn/ang/booking/model/entities/Booking.java
package learn.ang.booking.model.entities;

import learn.ang.booking.model.Layout;
import lombok.Getter;
import lombok.Setter;

import javax.persistence.*;
import java.sql.Date;
import java.sql.Time;

@Entity
@Getter
@Setter
public class Booking {
    @Id
    @GeneratedValue(strategy = GenerationType.AUTO)
    private Long id;

    @ManyToOne
    private Room room;

    @ManyToOne
    private User user;

    @Enumerated(EnumType.STRING)
    private Layout layout;

    private String title;
    private Date date;
    private Time startTime;
    private Time endTime;
    private Integer participants;

    public Booking() {
    }

    public Booking(Room room, User user, Layout layout, String title, Date date, Time startTime, Time endTime, Integer participants) {
        this.room = room;
        this.user = user;
        this.layout = layout;
        this.title = title;
        this.date = date;
        this.startTime = startTime;
        this.endTime = endTime;
        this.participants = participants;
    }

    public Integer getDisplayDay() {
        return this.date.toLocalDate().getDayOfMonth();
    }

    public String getDisplayMonth() {
        return this.date.toLocalDate().getMonth().toString();
    }

    public String getDisplayDayOfWeek() {
        return this.date.toLocalDate().getDayOfWeek().toString();
    }
}

File: BookingClient/src/app/model/Booking.ts
import {Layout, Room} from "./Room";
import {User} from "./User";

export class Booking {
  // @ts-ignore
  id: number;
  room: Room;
  user: User;
  layout: Layout;
  title: string;
  date: string;
  startTime: string;
  endTime: string;
  participants: number;


  constructor(room?: Room, user?: User, layout?: Layout, title?: string, date?: string,
              startTime?: string, endTime?: string, participants?: number, id?: number) {
    if (room) {
      this.room = room;
    }
    if (user) {
      this.user = user;
    }
    if (layout) {
      this.layout = layout;
    }
    if (title) {
      this.title = title;
    }
    if (date) {
      this.date = date;
    }
    if (startTime) {
      this.startTime = startTime;
    }
    if (endTime) {
      this.endTime = endTime;
    }
    if (participants) {
      this.participants = participants;
    }
    if (id) {
      this.id = id;
    }
  }

  public copyTo(other: Booking): Booking {
    other.room = this.room;
    other.user = this.user;
    other.layout = this.layout;
    other.title = this.title;
    other.date = this.date;
    other.startTime = this.startTime;
    other.endTime = this.endTime;
    other.participants = this.participants;
    return other;
  }

  public getDateAsDate(): Date {
    return new Date(this.date);
  }

  public isNew(): boolean {
    return this.id == null;
  }

  static fromHttp(jsBooking: Booking): Booking {
    const newBooking = new Booking();
    newBooking.id = jsBooking.id;
    newBooking.room = Room.fromHttp(jsBooking.room);
    newBooking.user = User.fromHttp(jsBooking.user);
    newBooking.layout = Layout[jsBooking.layout];
    newBooking.title = jsBooking.title;
    newBooking.date = jsBooking.date;
    newBooking.startTime = jsBooking.startTime;
    newBooking.endTime = jsBooking.endTime;
    newBooking.participants = jsBooking.participants;
    return newBooking;
  }
}
